### PR TITLE
For legacy inpaint pipeline, scale mask tensor to latent size and avoid noise added to final latents

### PR DESCRIPTION
**Pull request by 1lint**
Follow up to https://github.com/huggingface/diffusers/pull/2067, does not include part 3).

PR addresses the following issues in the stable diffusion legacy inpaint pipeline:
1. When passing the inpainting mask as a tensor, the pipeline does not scale the mask to the latent size, causing a shape mismatch during the denoising loop. 
Fix: Added logic to resize the mask tensor to the latent size with torch interpolate.

2. Noise is added to the unmasked latents after the denoising loop, causing noticeable image corruption with certain schedulers/low inference steps
Fix: Restore original unmasked latents after denoising loop.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py [hunk 1]
```diff
@@ -562,6 +567,9 @@ def __call__(
                     if callback is not None and i % callback_steps == 0:
                         callback(i, t, latents)
 
+        # use original latents corresponding to unmasked portions of the image
+        latents = (init_latents_orig * mask) + (latents * (1 - mask))
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:571]: Hmm this changes the behavior quite a bit of the legacy pipeline, but it does make sense! @patil-suraj @pcuenca wdyt?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py [hunk 2]
```diff
@@ -509,6 +509,11 @@ def __call__(
 
         if not isinstance(mask_image, torch.FloatTensor):
             mask_image = preprocess_mask(mask_image, self.vae_scale_factor)
+        else:
+            # permute dimensions to pytorch standard (B, C, H, W)
+            mask_image = mask_image.permute(0, 3, 1, 2)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:514]: Thanks! Could we add a test here?

**Review by patrickvonplaten — COMMENTED** [review]
Thanks for making the PR simpler :-) 

A bit worried since the changes affect every image generation, but they seem sensible to me! 

@pcuenca @patil-suraj what do you think?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py (hunk 2 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:514]: I think we should do some additional checks here. What if `maks_image` is already `channels_first` ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py (hunk 1 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:571]: Makes sense!

**Review by patil-suraj — COMMENTED** [review]
The changes make sense to me, I just left a comment about permuting `mask_image`, as we accept `tensors`, it could already be in the right format.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py (hunk 2 above)
Comment (1lint) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint_legacy.py:514]: Good point, user may pass mask channel first in standard shape. I'm thinking of the test below for expected mask input mask shape (B x 1 x H x W) or (B x H x W x 1). 
Also worth considering, do you want to support user passing mask tensor with 3 channel dimensions? (B x 3 x H x W)?

`import torch

image = torch.randn(1,3,512,512)
mask = torch.randn(1,512,512,1)

# squeeze potential singleton channel dimensions, then unsqueeze to pytorch standard (B x 1 x H x W)
mask = mask.squeeze(3).squeeze(1).unsqueeze(1)

if mask.shape[-2:] != image.shape[-2:]:
    raise ValueError(
            f"Mask height and width {tuple(mask.shape[-2:])} should match image height and width {tuple(image.shape[-2:])}"
        )
mask.shape`
